core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/AbstractHealthCheckedEndpointGroupBuilder.java

```diff
@@ -368,6 +379,52 @@ public SELF selectionTimeoutMillis(long initialSelectionTimeoutMillis, long sele
         return self();
     }
 
+    /**
+     * Sets a predicate to filter health checked {@link Endpoint}s. Whenever there is an update
+     * in endpoints, this predicate is used to filter {@link Endpoint}s that should be considered
+     * for selection.
+     *
+     * <p>For example:<pre>{@code
+     * // regardless of health check status, all endpoints will be considered for selection
+     * HealthCheckedEndpointGroup endpointGroup =
+     *     HealthCheckedEndpointGroup.builder(delegate, "/health")
+     *                               .healthCheckedEndpointPredicate(endpoint -> true)
+     *                               .build();
+     * }</pre>
+     */
+    @UnstableApi
+    public SELF healthCheckedEndpointPredicate(Predicate<Endpoint> healthCheckedEndpointPredicate) {
+        this.healthCheckedEndpointPredicate =
+                requireNonNull(healthCheckedEndpointPredicate, "healthCheckedEndpointPredicate");
+        return self();
+    }
+
+    /**
+     * Sets a predicate to determine whether a list of {@link Endpoint} should be considered healthy
+     * before the initial health check. This predicate will be called on each update of the delegate
+     * {@link EndpointGroup} as long as the first round of health checks haven't completed yet.
+     * This can be used to potentially avoid a long initial endpoint selection time. If an empty list
+     * is returned, the {@link HealthCheckedEndpointGroup} waits until the first round of health checks
+     * complete.
+     *
+     * <p>For example:<pre>{@code
+     * HealthCheckedEndpointGroup endpointGroup =
+     *     HealthCheckedEndpointGroup.builder(delegate, "/health")
+     *                               // all endpoints are candidates for endpoint selection
+     *                               .initialStateResolver(endpoints -> endpoints)
+     *                               // we wait until the first round of health checks complete
+     *                               // before considering endpoints for selection
+     *                               .initialStateResolver(endpoints -> Collections.emptyList())
+     *                               .build();
+     * }</pre>
+     */
+    @UnstableApi
+    public SELF initialStateResolver(
+            Function<? super List<Endpoint>, ? extends List<Endpoint>> initialStateResolver) {
```

Comment:
I see, since the xDS implementation needs the health check attribute to be retained from filtered `initialStateResolver`,  an `AttributeRetainingEndpointGroup` has been introduced

Reply:
Spent way too long debugging an issue related to this. I realized that `initialStateResolver` can have a race condition with the first update of healthy candidates. As much as I dislike flags, I ended up adding a `firstRun` boolean flag instead of using the `initialized` flag in `HealthCheckedEndpointGroup`.